presto-orc/src/main/java/com/facebook/presto/orc/writer/FlatMapKeyNodeWriter.java

```diff
@@ -0,0 +1,170 @@
+    private final CompressedMetadataWriter metadataWriter;
+    private final List<ColumnStatistics> rowGroupColumnStatistics = new ArrayList<>();
+    private boolean closed;
+    private int nonNullValueCount;
+
+    public int getSequence()
+    {
+        return this.sequence;
+    }
+
+    public FlatMapKeyNodeWriter(int column, int sequence, ColumnWriterOptions columnWriterOptions, Optional<DwrfDataEncryptor> dwrfEncryptor, MetadataWriter metadataWriter, ColumnWriter valueWriter)
+    {
+        checkArgument(column >= 0, "column is negative");
+        checkArgument(sequence >= 1, "sequence is non positive");
+        this.column = column;
+        this.sequence = sequence;
+        this.valueWriter = valueWriter;
+        inMapStream = new BooleanOutputStream(columnWriterOptions, dwrfEncryptor);
+        this.metadataWriter = new CompressedMetadataWriter(metadataWriter, columnWriterOptions, dwrfEncryptor);
+    }
+
+    @Override
+    public List<ColumnWriter> getNestedColumnWriters()
+    {
+        return valueWriter.getNestedColumnWriters();
+    }
+
+    @Override
+    public Map<Integer, ColumnEncoding> getColumnEncodings()
+    {
+        Map<Integer, ColumnEncoding> valueEncoding = valueWriter.getColumnEncodings();
+        ImmutableMap.Builder<Integer, ColumnEncoding> resultEncoding = ImmutableMap.builder();
+        for (Map.Entry<Integer, ColumnEncoding> encoding : valueEncoding.entrySet()) {
+            resultEncoding.put(this.sequence, encoding.getValue());
+        }
+        return resultEncoding.build();
+    }
+
+    @Override
+    public void beginRowGroup()
+    {
+        valueWriter.beginRowGroup();
+        inMapStream.recordCheckpoint();
+    }
+
+    public void addNulls(int count)
+    {
+        this.inMapStream.writeBooleans(count, false);
+    }
+
+    @Override
+    public void writeBlock(Block block)
+    {
+        this.nonNullValueCount++;
+        this.inMapStream.writeBoolean(true);
+        valueWriter.writeBlock(block);
+    }
+
+    @Override
+    public Map<Integer, ColumnStatistics> finishRowGroup()
+    {
+        return valueWriter.finishRowGroup();
+    }
+
+    @Override
+    public void close()
+    {
+        closed = true;
+        valueWriter.close();
+        inMapStream.close();
+    }
+
+    @Override
+    public Map<Integer, ColumnStatistics> getColumnStripeStatistics()
+    {
+        checkState(closed);
+        ImmutableMap.Builder<Integer, ColumnStatistics> columnStatistics = ImmutableMap.builder();
+        columnStatistics.putAll(valueWriter.getColumnStripeStatistics());
+        return columnStatistics.build();
+    }
+
+    @Override
+    public List<StreamDataOutput> getIndexStreams() throws IOException
+    {
+        checkState(closed);
+        /* Each Sequence has one ROW INDEX associated with the writer */
+        ImmutableList.Builder<StreamDataOutput> indexStreams = ImmutableList.builder();
+        indexStreams.addAll(valueWriter.getIndexStreams());
```

Comment:
Use ImmutableList.copyOf instead. Do we really need to make them immutable, looks like getIndexStreams is already doing that.